File: algos/algos-1.0.0/ts/algorithms/strings/is_palindrome/is_palindrome.ts
export function isPalindrome(input: string) {
  let start = 0;
  let end = input.length - 1;

  while (start < end) {
    if (input[start] === input[end]) {
      start++;
      end--;
    } else {
      return false;
    }
  }
  return true;
}

export function isPalindromeRecursive(input: string) {
  if (input.length === 1) {
    return true;
  }

  if (input.length === 2) {
    return input[0] === input[1];
  }

  if (input[0] === input[input.length - 1]) {
    return isPalindromeRecursive(input.slice(1, input.length - 1));
  }

  return false;
}

let res = isPalindromeRecursive("abba");
console.log("res", res);

File: algos/algorithms/rs/src/algorithms/strings/is_palindrome.rs
#[allow(dead_code)]
pub fn is_palindrome(input: &str) -> bool {
    let mut start: usize = 0;
    let mut end = input.len() - 1;
    while start < end {
        if input.chars().nth(start).unwrap() == input.chars().nth(end).unwrap() {
            start += 1;
            end -= 1;
        } else {
            return false;
        }
    }
    true
}

#[allow(dead_code)]
fn is_palindrome2(input: &str) -> bool {
    return input.chars().rev().collect::<String>() == input;
}

#[cfg(test)]
mod tests {
    use super::is_palindrome;
    #[test]
    fn test_is_palindrome() {
        assert_eq!(is_palindrome("racecar"), true);
        assert_eq!(is_palindrome("marcell"), false);
    }

    use super::is_palindrome2;
    #[test]
    fn test_is_palindrome2() {
        assert_eq!(is_palindrome2("racecar"), true);
        assert_eq!(is_palindrome2("marcell"), false);
    }
}
